### PR TITLE
feat: Use local decrypt/reencrypt

### PR DESCRIPTION
**Pull request by amaury1093**
Switch to use local decrypt/reencrypt call to `tfhe-rs` instead of doing a gRPC call to a KMS.

New code is compared to https://github.com/zama-ai/go-ethereum/blob/1.10.19-zama/core/vm/contracts.go

Tested locally, and it worked well deploying the IncoSlots contract based on the new 0.8.20 TFHE library. Here are the node logs:
```
Info: delegateCiphertextHandlesInArgs handle 0x681938caace4aa448d4e9c8822d595a4c776ce30fc8ec82d3f1752a0d3405214 fromDepth 1 toDepth 2 
Info: decrypt success plaintext 65153 
Info: Run removing ciphertext from depth handle 0x681938caace4aa448d4e9c8822d595a4c776ce30fc8ec82d3f1752a0d3405214 depth 1 
12:13PM INF executed block height=904 module=state num_invalid_txs=0 num_valid_txs=1 server=node
```

TODO:
- [x] Make sure tests pass locally
- [x] Make sure tests are run on Github Actions

### CODE REVIEW

File: fhevm/precompiles.go [hunk 1]
```diff
@@ -2069,58 +2064,25 @@ func reencryptRun(environment EVMEnvironment, caller common.Address, addr common
...
-			fheType = kms.FheType_Euint8
-		case FheUint16:
-			fheType = kms.FheType_Euint16
-		case FheUint32:
-			fheType = kms.FheType_Euint32
+		optReqResult, optReqErr := evaluateRemainingOptimisticRequires(environment)
+		if optReqErr != nil {
+			return nil, optReqErr
+		} else if !optReqResult {
+			return nil, ErrExecutionReverted
```
Comment (amaury1093) [inline fhevm/precompiles.go:2071]: These 4 lines are commented in Zama's new fhevm-go, but not commented in their old go-ethereum fork [here](https://github.com/zama-ai/go-ethereum/blob/f24d3e742ec643b848e69b6daab671dc787f7d5d/core/vm/contracts.go#L2053-L2058).

Any idea if we should keep them?

File: fhevm/precompiles.go [hunk 2]
```diff
@@ -2205,49 +2167,8 @@ func decryptRun(environment EVMEnvironment, caller common.Address, addr common.A
...
-
-	res, err := ep.Decrypt(ctx, decryptionRequest)
-	if err != nil {
-		logger.Error("decrypt failed", "err", err)
-		return 0, err
-	}
-
-	return uint64(res.Plaintext), err
+	v, err := ct.decrypt()
+	return v.Uint64(), err
```
Comment (amaury1093) [inline fhevm/precompiles.go:2171]: copy paste of https://github.com/zama-ai/go-ethereum/blob/f24d3e742ec643b848e69b6daab671dc787f7d5d/core/vm/contracts.go#L2188-L2203, but removing everything related to oracles

File: fhevm/tfhe.go [hunk 3]
```diff
@@ -1609,12 +1609,19 @@ func InitGlobalKeysFromFiles(keysDir string) error {
...
+	if err != nil {
+		return err
+	}
 
 	sks = C.deserialize_server_key(toBufferView(sksBytes))
 
 	pksHash = crypto.Keccak256Hash(pksBytes)
 	pks = C.deserialize_compact_public_key(toBufferView(pksBytes))
 
+	cks = C.deserialize_client_key(toBufferView(cksBytes))
```
Comment (amaury1093) [inline fhevm/tfhe.go]: copy-paste from https://github.com/zama-ai/go-ethereum/blob/f24d3e742ec643b848e69b6daab671dc787f7d5d/core/vm/tfhe.go#L1539-L1544

File: fhevm/precompiles.go (hunk 1 above)
Comment (lovenoble) [reply fhevm/precompiles.go:2071]: I don't think it would be a bad thing to keep it.

**Pull request merged**